### ISSUE TITLE
Update Google Maps definitions to latest version

### ISSUE THREAD

**Issue by cgwrench** [borisyankov/DefinitelyTyped#4364]
The latest release version of the Google Maps JavaScript API is v3.20, and was released on 21 May 2015. The current Google Maps definitions, see [googlemaps/google.maps.d.ts](https://github.com/borisyankov/DefinitelyTyped/blob/7d2f0aa875cc3bc41f3567e3ff99d344df986155/googlemaps/google.maps.d.ts), do not match [the latest version of the API](https://developers.google.com/maps/documentation/javascript/reference).

Examples of where the definitions are incorrect include:

- Definition of `LatLngLiteral` is missing. This affects numerous methods. We should define `export type LatLngLiteral = { lat: number; lng: number };`
- The Data layer API is missing (see #2249)
- `MarkerImage` is defined, but is no longer part of the API
- The `Marker` class `getIcon` and `setIcon` methods should return and accept `string`, `Icon` or `Symbol` objects, not `string` or `MarkerImage` objects
- The `Marker` class has `getFlat` and `getShadow` methods which are no longer part of the API
- The `Marker`  class is missing the methods `getAttribution`, `getOpacity`, `getPlace` (and the corresponding `set*` methods)
- The `KmlLayer` class is missing the methods `getZIndex`, `setUrl` and `setZIndex`.

Note that the above list is not exhaustive.

**Comment by vvakame** [comment]
@cgwrench thanks your pull request!
please make additional pull request if you can... :wink:

**Comment by cgwrench** [comment]
@vvakame I have made a start on this. However, this is quite a large library so will take a bit of time for me to update completely. 

The following is a list of all types defined in the latest [JavaScript API reference](https://developers.google.com/maps/documentation/javascript/reference). Those marked as done are the ones which I have updated so far.

- Map

  - [X] Map
  - [X] MapOptions
  - [X] MapTypeId

- Controls

  - [X] MapTypeControlOptions
  - [X] MapTypeControlStyle
  - [X] OverviewMapControlOptions
  - [X] PanControlOptions
  - [X] RotateControlOptions
  - [X] ScaleControlOptions
  - [X] ScaleControlStyle
  - [X] StreetViewControlOptions
  - [X] ZoomControlOptions
  - [X] ZoomControlStyle
  - [X] ControlPosition

- Data

  - [X] Data
  - [X] Data.DataOptions
  - [X] Data.GeoJsonOptions
  - [X] Data.StyleOptions
  - [X] Data.StylingFunction
  - [X] Data.Feature
  - [X] Data.FeatureOptions
  - [X] Data.Geometry
  - [X] Data.Point
  - [X] Data.MultiPoint
  - [X] Data.LineString
  - [X] Data.MultiLineString
  - [X] Data.LinearRing
  - [X] Data.Polygon
  - [X] Data.MultiPolygon
  - [X] Data.GeometryCollection
  - [X] Data.MouseEvent
  - [X] Data.AddFeatureEvent
  - [X] Data.RemoveFeatureEvent
  - [X] Data.SetGeometryEvent
  - [X] Data.SetPropertyEvent
  - [X] Data.RemovePropertyEvent
  
- Overlays

  - [X] Marker
  - [X] MarkerOptions
  - [X] Icon
  - [X] MarkerShape
  - [X] Symbol
  - [X] SymbolPath
  - [X] Animation
  - [X] InfoWindow
  - [X] InfoWindowOptions
  - [X] Polyline
  - [X] PolylineOptions
  - [X] IconSequence
  - [X] Polygon
  - [X] PolygonOptions
  - [X] PolyMouseEvent
  - [X] Rectangle
  - [X] RectangleOptions
  - [X] Circle
  - [X] CircleOptions
  - [X] StrokePosition
  - [X] GroundOverlay
  - [X] GroundOverlayOptions
  - [X] OverlayView
  - [X] MapPanes
  - [X] MapCanvasProjection

- Services
 
  - [X] Geocoder
  - [X] GeocoderRequest
  - [X] GeocoderComponentRestrictions
  - [X] GeocoderStatus
  - [X] GeocoderResult
  - [X] GeocoderAddressComponent
  - [X] GeocoderGeometry
  - [X] GeocoderLocationType
  - [X] DirectionsRenderer
  - [X] DirectionsRendererOptions
  - [X] DirectionsService
  - [X] DirectionsRequest
  - [X] TravelMode
  - [X] UnitSystem
  - [X] TransitOptions
  - [X] TransitMode
  - [X] TransitRoutePreference
  - [X] TransitFare
  - [X] DirectionsWaypoint
  - [X] DirectionsStatus
  - [X] DirectionsResult
  - [X] DirectionsRoute
  - [X] DirectionsLeg
  - [X] DirectionsStep
  - [X] Distance
  - [X] Duration
  - [X] Time
  - [X] TransitDetails
  - [X] TransitStop
  - [X] TransitLine
  - [X] TransitAgency
  - [X] TransitVehicle
  - [X] VehicleType
  - [X] ElevationService
  - [X] LocationElevationRequest
  - [X] PathElevationRequest
  - [X] ElevationResult
  - [X] ElevationStatus
  - [X] MaxZoomService
  - [X] MaxZoomResult
  - [X] MaxZoomStatus
  - [X] DistanceMatrixService
  - [X] DistanceMatrixRequest
  - [X] DistanceMatrixResponse
  - [X] DistanceMatrixResponseRow
  - [X] DistanceMatrixResponseElement
  - [X] DistanceMatrixStatus
  - [X] DistanceMatrixElementStatus

- Save to Google Maps

  - [X] Attribution
  - [X] Place
  - [X] SaveWidget
  - [X] SaveWidgetOptions

- Map Types

  - [ ] MapType
  - [ ] MapTypeRegistry
  - [ ] Projection
  - [ ] ImageMapType
  - [ ] ImageMapTypeOptions
  - [ ] StyledMapType
  - [ ] StyledMapTypeOptions
  - [ ] MapTypeStyle
  - [ ] MapTypeStyleFeatureType
  - [ ] MapTypeStyleElementType
  - [ ] MapTypeStyler

- Layers

  - [ ] BicyclingLayer
  - [ ] FusionTablesLayer
  - [ ] FusionTablesLayerOptions
  - [ ] FusionTablesQuery
  - [ ] FusionTablesStyle
  - [ ] FusionTablesHeatmap
  - [ ] FusionTablesMarkerOptions
  - [ ] FusionTablesPolygonOptions
  - [ ] FusionTablesPolylineOptions
  - [ ] FusionTablesMouseEvent
  - [ ] FusionTablesCell
  - [ ] KmlLayer
  - [ ] KmlLayerOptions
  - [ ] KmlLayerMetadata
  - [ ] KmlLayerStatus
  - [ ] KmlMouseEvent
  - [ ] KmlFeatureData
  - [ ] KmlAuthor
  - [ ] TrafficLayer
  - [ ] TransitLayer
  
- Street View

  - [ ] StreetViewPanorama
  - [ ] StreetViewPanoramaOptions
  - [ ] StreetViewAddressControlOptions
  - [ ] StreetViewLink
  - [ ] StreetViewPov
  - [ ] StreetViewPanoramaData
  - [ ] StreetViewLocation
  - [ ] StreetViewTileData
  - [ ] StreetViewService
  - [ ] StreetViewStatus
  - [ ] StreetViewCoverageLayer

- Events

  - [ ] MapsEventListener
  - [ ] event
  - [ ] MouseEvent

- Base

  - [X] LatLng
  - [X] LatLngLiteral
  - [X] LatLngBounds
  - [X] Point
  - [X] Size

- MVC

  - [X] MVCObject
  - [X] MVCArray

- Geometry Library

  - [ ] encoding
  - [ ] spherical
  - [ ] poly

- AdSense Library

  - [ ] AdUnit
  - [ ] AdUnitOptions
  - [ ] AdFormat

- Panoramio Library

  - [ ] PanoramioLayer
  - [ ] PanoramioLayerOptions
  - [ ] PanoramioFeature
  - [ ] PanoramioMouseEvent

- Places Library

  - [ ] Autocomplete
  - [ ] AutocompleteOptions
  - [ ] AutocompletePrediction
  - [ ] PredictionTerm
  - [ ] PredictionSubstring
  - [ ] AutocompleteService
  - [ ] AutocompletionRequest
  - [ ] ComponentRestrictions
  - [ ] PlaceAspectRating
  - [ ] PlaceDetailsRequest
  - [ ] PlaceGeometry
  - [ ] PlacePhoto
  - [ ] PhotoOptions
  - [ ] PlaceResult
  - [ ] PlaceReview
  - [ ] PlaceSearchPagination
  - [ ] PlaceSearchRequest
  - [ ] PlacesService
  - [ ] PlacesServiceStatus
  - [ ] QueryAutocompletePrediction
  - [ ] QueryAutocompletionRequest
  - [ ] RadarSearchRequest
  - [ ] RankBy
  - [ ] SearchBox
  - [ ] SearchBoxOptions
  - [ ] TextSearchRequest

- Drawing Library

  - [ ] DrawingManager
  - [ ] DrawingManagerOptions
  - [ ] DrawingControlOptions
  - [ ] OverlayCompleteEvent
  - [ ] OverlayType

- Weather Library

  - [ ] CloudLayer
  - [ ] WeatherLayer
  - [ ] WeatherLayerOptions
  - [ ] TemperatureUnit
  - [ ] WindSpeedUnit
  - [ ] LabelColor
  - [ ] WeatherMouseEvent
  - [ ] WeatherFeature
  - [ ] WeatherConditions
  - [ ] WeatherForecast

- Visualization Library

  - [ ] MapsEngineLayer
  - [ ] MapsEngineLayerOptions
  - [ ] MapsEngineLayerProperties
  - [ ] MapsEngineMouseEvent
  - [ ] MapsEngineStatus
  - [ ] DynamicMapsEngineLayer
  - [ ] DynamicMapsEngineLayerOptions
  - [ ] DynamicMapsEngineMouseEvent
  - [ ] FeatureStyle
  - [ ] HeatmapLayer
  - [ ] HeatmapLayerOptions
  - [ ] WeightedLocation

# Outstanding issues

- `Element` is used in the definition in multiple places. The API documentation uses `Node`. The definitions should be updated to use `Node` instead of `Element`.

- A number of parts of the API reference a typed `MVCArray`, e.g. `MVCArray<LatLng>`. Should a generic `MVCArray<T>` be added to the definitions.

- A number of parts of the API reference a type of  `Array<LatLng|LatLngLiteral>`. Is this accurately reflected by using `LatLng[]|LatLngLiteral[]`?

**Comment by rayshan** [comment]
+1

**Comment by PrashantAshok** [comment]
Hey @cgwrench , it looks like InfoWindow.open() has been incorrectly updated in this pull request, as per [3.20](https://developers.google.com/maps/documentation/javascript/reference#InfoWindow) version it still accepts `Map` as option. 

```
open(map?:Map|StreetViewPanorama, anchor?:MVCObject)
```

Can you please fix it?

**Comment by cgwrench** [comment]
@PrashantAshok Good spot. I did initialy change this to be a single function definition as part of 3e467db, however, this appears to have been overwritten by a subsequent merge.

This issue should be fixed by #4913.

**Comment by PrashantAshok** [comment]
Thanks @cgwrench for the quick fix. I also noticed that all params of GeocoderComponentRestrictions should be optional as per the [spec](https://developers.google.com/maps/documentation/javascript/reference#GeocoderComponentRestrictions)

```
export interface GeocoderComponentRestrictions {
        administrativeArea: string;
        country: string;
        locality: string;
        postalCode:	string;
        route: string;
}
```

btw, let me know if you need any help in updating type def for any remaining items, I can try to pitch in - places library?

**Comment by cgwrench** [comment]
@PrashantAshok Thanks for raising this. This issue should be fixed by #4993. I've tried to check all options objects for the correct specification of optional properties, but I may have missed some. Feel free to raise these as new issues if you spot any.

The entire library should be up to date (with version 3.20 of the Google Maps API). However, the tests are severly lacking. If you would like to contribute some tests that would be really helpful. Feel free to start with a single aspect of the API, such as the Places library.

**Comment by mikedon** [comment]
For the GeocoderRequest interface doesn't there still need to be a "latLng" property?  I just noticed this because we use the reverse lookup api

https://developers.google.com/maps/documentation/geocoding/intro#ReverseGeocoding
